test/util/memoizer-test.ts

```diff
@@ -0,0 +1,444 @@
+    const d1 = new Date(2000, 1, 1);
+    const d2 = new Date(2000, 1, 1);
+    const d3 = new Date(2000, 1, 1, 1);
+
+    const id1 = immutableMemoizer.toImmutableObjectKey(d1);
+    const id2 = immutableMemoizer.toImmutableObjectKey(d2);
+    const id3 = immutableMemoizer.toImmutableObjectKey(d3);
+
+    // Classic
+    equals(d1, d2).should.be.true();
+    equals(d2, d1).should.be.true();
+    equals(d1, d3).should.be.false();
+    equals(d3, d1).should.be.false();
+
+    // Immutable
+    Immutable.is(id1, id2).should.be.true();
+    Immutable.is(id2, id1).should.be.true();
+    Immutable.is(id1, id3).should.be.false();
+    Immutable.is(id3, id1).should.be.false();
+  });
+
+  it('should properly match regular expressions', () => {
+    const r1 = /^\d*(\.\d{1,2})?$/g;
+    const r2 = /^\d*(\.\d{1,2})?$/g;
+    const r3 = /^([0-9]{0,5}-)?[0-9]{10}$/g;
+
+    const ir1 = immutableMemoizer.toImmutableObjectKey(r1);
+    const ir2 = immutableMemoizer.toImmutableObjectKey(r2);
+    const ir3 = immutableMemoizer.toImmutableObjectKey(r3);
+
+    // Classic
+    equals(r1, r2).should.be.true();
+    equals(r2, r1).should.be.true();
+    equals(r1, r3).should.be.false();
+    equals(r3, r1).should.be.false();
+
+    // Immutable
+    Immutable.is(ir1, ir2).should.be.true();
+    Immutable.is(ir2, ir1).should.be.true();
+    Immutable.is(ir1, ir3).should.be.false();
+    Immutable.is(ir3, ir1).should.be.false();
+  });
+
+  it('should properly match arrays', () => {
+    const c1 = [1, null];
+    const c2 = [1, null];
+    const c3 = [1, 2];
+
+    const ic1 = immutableMemoizer.toImmutableObjectKey(c1);
+    const ic2 = immutableMemoizer.toImmutableObjectKey(c2);
+    const ic3 = immutableMemoizer.toImmutableObjectKey(c3);
+
+    // Classic behavior
+    should(equals(c1, c2)).be.null();
+    should(equals(c1, c3)).be.null();
+
+    // The classic behavior is incorrect for List except, intersect, and distinct operations
+    // For each, null should be handled as equals according to spec.  Here's the quote from CQL Appendix B:
+    // Distinct: The operator uses equality comparison semantics as defined in the Equal operator, with the exception that nulls are considered equal for the purposes of distinct determination. This means that multiple nulls in the input will result in a single null in the output.
+    // Except: This operator uses equality semantics to determine whether two elements are the same for the purposes of computing the difference, with the exception that null elements are considered equal.
+    // Intersect: This operator uses equality semantics to determine whether or not two elements are the same, with the exception that null elements are considered equal for the purposes of the intersection.
+    // A CQL example that illustrates the incorrect behavior of old code for the distinct operator, and correct behavior of new code is:
+    // define "Example":
+    //     ({ {1, null}, {1, null} }) q
+    //     return q
```

Comment:
This example doesn't quite demonstrate what you discuss above because the two arguments you are doing `distinct` on are lists themselves (not `null`). Since they are lists, equality semantics for comparing two lists apply.  But as it happens, that also has an exception for nulls (which is why your solution is still correct):
> For list types, this means that equality returns true if and only if the lists contain elements of the same type, have the same number of elements, and for each element in the lists, in order, the elements are equal using equality semantics, with the exception that null elements are considered equal.

Anyway, a better example if you're trying to show that `distinct` collapses nulls would be:
```
define "Example":
    ({ null, null }) q
    return q
```
But I think this one actually did work in "classic" mode.  So it's really equality of two lists, each of which have nulls at the same indices, that maybe was broken in classic mode (and maybe still is).

Reply:
> But I think this one actually did work in "classic" mode. So it's really equality of two lists, each of which have nulls at the same indices, that maybe was broken in classic mode (and maybe still is).

Yup - this is right.  Collapsing nulls worked as expected.  Collapsing equal lists (with null) is the problem.  This was broken in classic mode.  The Immutable.is implementation deviates from the classic behavior and corrects this problem.